### PR TITLE
Fix cudnn channels_last descriptors problem

### PR DESCRIPTION
**Pull request by jjsjann123**
This is to append fixes to #31783 so we can pull the fixes in without breaking tests.

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 86854658:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=86854658f2a97ab0fad86a4a7937e495761afaca).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4169763) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=81935669) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=81935669))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
 
HEAD is now at 86854658f2 fixing suggest_memory_format 
+ git reset --hard 86854658f2a97ab0fad86a4a7937e495761afaca 
HEAD is now at 86854658f2 fixing suggest_memory_format 
+ git merge --allow-unrelated-histories --no-edit --no-ff 7f723cbd8a53c85722af14d8749559de1992cd9b 
Auto-merging test/test_nn.py 
Removing aten/src/THNN/generic/RReLU.c 
Auto-merging aten/src/ATen/native/cudnn/Conv.cpp 
CONFLICT (content): Merge conflict in aten/src/ATen/native/cudnn/Conv.cpp 
Auto-merging aten/src/ATen/native/Convolution.cpp 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: aten/src/ATen/templates/TensorBody.h [hunk 1]
```diff
@@ -204,9 +204,17 @@ class CAFFE2_API Tensor {
     return impl_->is_non_overlapping_and_dense();
   }
 
-  at::MemoryFormat suggest_memory_format() const {
-    if (!is_mkldnn() && !is_sparse() && !impl_->is_contiguous() && impl_->is_strides_like_channels_last()) {
-      return at::MemoryFormat::ChannelsLast;
+  at::MemoryFormat suggest_memory_format(
```
Comment (jjsjann123) [inline aten/src/ATen/templates/TensorBody.h]: I'm not quite uncertain that I understand the behavior of `suggest_memory_format` here. I would prefer to do this right in a separate PR (add tests to define the proper behavior), so hopefully when Vitaly comes back it would be easier to communicate it through.

**Comment by jjsjann123** [comment]
I'll keep an eye on the tests as they run. Otherwise, this is good for review.

**Comment by jjsjann123** [comment]
How does script decorator stack?

Vitaly's code put cudnn version check after `TEST_CUDNN`, I think it means that cudnn is guaranteed to be around?
https://github.com/pytorch/pytorch/blob/52e07eb39a1cdfe142b2b5f904e598e64f22a8ee/test/test_nn.py#L7588-L7590

If this is not supposed to work, I can do it as our `TEST_LARGE_TENSOR` thing: https://github.com/pytorch/pytorch/blob/52e07eb39a1cdfe142b2b5f904e598e64f22a8ee/test/test_nn.py#L62-L64

**Comment by ngimel** [comment]
Please write the new tests using device-generic framework https://github.com/pytorch/pytorch/wiki/Writing-tests-that-run-on-all-available-device-types, it also has a decorator that would run with specified cudnn versions only, @skipCUDAIfCudnnVersionLessThan (in common_device_type.py)

**Comment by jjsjann123** [comment]
Tests updated.

**Comment by jjsjann123** [comment]
Failing test seems to be irrelevant?

File: aten/src/ATen/cudnn/Descriptors.h [hunk 2]
```diff
@@ -35,6 +35,13 @@ inline int dataSize(cudnnDataType_t dataType)
 // function modifies 'stride' in place so this invariant holds.
 static inline void fixSizeOneDimStride(int dim, const int *size, int *stride) {
   int64_t z = 1;
+  if (dim == 4) {
```
Comment (ngimel) [inline aten/src/ATen/cudnn/Descriptors.h]: Is this function used anywhere? Logic seems fishy to me - if nhwc-contiguous with already good 1-dim strides, bail out, but if say these strides are messed up, they are going to get messed up even more below.

File: aten/src/ATen/native/cudnn/Conv.cpp [hunk 3]
```diff
@@ -125,13 +125,16 @@ Tensor narrowGroup(const Tensor& t, int dim, int group_idx, int64_t groups) {
   return t.narrow(dim, group_idx * group_size, group_size);
 }
 
-bool support_channels_last(const Tensor& input, const Tensor& weight, int groups) {
+bool support_channels_last(
+    const Tensor& input,
+    const Tensor& weight,
```
Comment (ngimel) [inline aten/src/ATen/native/cudnn/Conv.cpp]: I don't see `weight` argument being used in the function

File: c10/core/TensorImpl.cpp [hunk 4]
```diff
@@ -116,14 +116,20 @@ bool TensorImpl::compute_channels_last_contiguous() const {
...
-        }
+      if (strides_[d] < min) {
+        return false;
+      }
+      // special case for N111, so we have NCHW as default layout;
+      // There's no way to disambiguate the memory format: issue #24090
+      if (d==0 && min==1) {
+        return false;
       }
+      min = std::max(strides_[d], sizes_[d]);
```
Comment (ngimel) [inline c10/core/TensorImpl.cpp]: a comment about how this works will be useful, maybe with some relevant examples

File: c10/core/TensorImpl.h [hunk 5]
```diff
@@ -1492,7 +1492,7 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
   void refresh_contiguous() {
     is_contiguous_ = compute_contiguous();
     is_channels_last_contiguous_ = compute_channels_last_contiguous();
-    is_channels_last_ = is_channels_last_contiguous_ || compute_strides_like_channels_last();
+    is_channels_last_ = compute_strides_like_channels_last();
```
Comment (ngimel) [inline c10/core/TensorImpl.h]: what happened here? It's always good to take shortcut if available.

File: aten/src/ATen/cudnn/Descriptors.h (hunk 2 above)
Comment (jjsjann123) [reply aten/src/ATen/cudnn/Descriptors.h]: Good catch. Forgot to update this guy after the last commit that fixed `suggest_memory_format()`. This function is somewhat busted for NHWC. But we actually always have good strides for NHWC before entering a convolution, so it's not relevant.

I'll update it.

Comment (jjsjann123) [reply aten/src/ATen/cudnn/Descriptors.h]: I mean, I'll properly fix this function so we would properly fix SizeOneDim for NHWC as well.

File: aten/src/ATen/native/cudnn/Conv.cpp (hunk 3 above)
Comment (jjsjann123) [reply aten/src/ATen/native/cudnn/Conv.cpp]: This is left for next PR. (Should rename this function to `use_channels_last` though). I want to update the behavior:
It be decided by the memory_format of the weight instead of input. (or maybe `input.suggest_memory_format() == ChannelsLast || weight.suggest_memory_format() == ChannelsLast`)

I'm OK if we want to get rid of weight here for now / or put a comment there.

File: c10/core/TensorImpl.h (hunk 5 above)
Comment (jjsjann123) [reply c10/core/TensorImpl.h]: Except this shortcut is not the right thing to do.

`is_channels_last_contiguous_` is not equivalent to `is_channels_last_`.
In ambiguous cases `NC11`, where the tensor has both `is_contiguous_` and `is_channels_last_contiguous_` set as true. We need to use `is_channels_last_` to represent the desired `memory_format`.

I'll put a comment here.

**Comment by jjsjann123** [comment]
To answer Natalia's request on benchmark.

`compute_strides_like_channels_last()` takes 5~6 ns on my machine;
Overall `refresh_contiguous()` takes about 20~40 ns.

While `compute_non_overlapping_and_dense()` is the big hammer and takes ~80ns. It doesn't run for common contiguous tensors, we should be fine.

File: c10/core/MemoryFormat.h [hunk 6]
```diff
@@ -60,4 +60,88 @@ inline std::vector<int64_t> get_channels_last_strides(IntArrayRef sizes) {
...
+      // 1. distinguish the memory_format of N1H1;
+      //     [H, 1, 1, 1] channels_last stride
+      //     [H, H, 1, 1] contiguous stride
+      // 2. permutation of 1C1W:
+      //     [1, C, 1, H]@[HC, H, H, 1] transpose(1, 3)
+      //     [1, H, 1, C]@[HC, 1, H, H] shouldn't be identified as channels_last
+      min = strides[d];
+      if (sizes[d] > 1) {
+        min *= (sizes[d]-1);
+      }
```
Comment (ngimel) [inline c10/core/MemoryFormat.h]: Yay, this is much better! Thank you for the comments!

**Review by ngimel — APPROVED** [review]
